### ISSUE TITLE
Train an active learning model during labeling

### ISSUE THREAD

**Issue by csaroff** [ryouchinsa/Rectlabel-support#143]
First of all, great tool!

In addition to automatically labeling images with Core ML, I was wondering if RectLabel could also train a model in the background with the existing labeled data in order to facilitate active learning.  I think that this could speed up the labeling process dramatically.

RectLabel could allow the user to order images by model confidence so that the user could target the images that are most confusing the model.

Thoughts?

**Comment by ryouchinsa** [comment]
Thanks for writing the issue.

At this moment, training the Core ML model is supported using the python script not on RectLabel.
https://github.com/ryouchinsa/Rectlabel-support/blob/master/train_turicreate.py

> I was wondering if RectLabel could also train a model in the background
If you have your own GPU, it might be comfortable.
But for almost users, running the training process in the background would take so much CPU power.

> I think that this could speed up the labeling process dramatically.
You might want to maintain the minimum dataset to achieve the high accuracy.

> RectLabel could allow the user to order images by model confidence
At this moment, our users are doing training and testing processes by themselves.
1. Label some images
2. Train the model
3. Test to unlabeled images
4. Label and add difficult images to training data

Let us know your opinion.

**Comment by csaroff** [comment]
> Thanks for writing the issue.

> At this moment, training the Core ML model is supported using the python script not on RectLabel.
> https://github.com/ryouchinsa/Rectlabel-support/blob/master/train_turicreate.py

I guess I didn't realize that it would be that simple :P

> If you have your own GPU, it might be comfortable.
> But for almost users, running the training process in the background would take so much CPU power.

It could be a feature that is off by default.  You wouldn't need to train the model from scratch every time an annotation is finished.  You could just retrain the model with the most recently labeled example.  Maybe you could train on a cpu if it was only one image at a time?

> At this moment, our users are doing training and testing processes by themselves.
>
> 1. Label some images
> 2. Train the model
> 3. Test to unlabeled images
> 4. Label and add difficult images to training data
>
> Let us know your opinion.

It would still be a nice feature for a number of reasons. 
1. Automating the training steps makes active learning accessible to a larger userbase. 
2. Training the model in the background saves users time.
3. Users will see if they have labeled enough examples to make a decent model without leaving the annotation tool.

If RectLabel supports this and allows model based data exploration from within the tool, it would provide a huge competitive advantage.

**Comment by ryouchinsa** [comment]
Thanks for the details.

You could write so many reasons to improve but we do not understand from which point we should improve to achieve the active learning.

For example, using the active learning, how are you training your current model without RectLabel?
We have to understand the details of the training flow.

> Maybe you could train on a cpu if it was only one image at a time?

If you add an image to the existing 100 images, you have to iterate training using batch sampling among the 101 images usually.

> If RectLabel supports this and allows model based data exploration from within the tool

What do you mean by "model based data exploration"?
Is it the image sorting using the confidence scores?

**Comment by csaroff** [comment]
> Thanks for the details.
> 
> You could write so many reasons to improve but we do not understand from which point we should improve to achieve the active learning.
>
> For example, using the active learning, how are you training your current model without RectLabel?
> We have to understand the details of the training flow.

So far I haven't used the active learning feature because I hadn't bothered to work through training a createml model, but I'll try it out soon.  The training script looks quite friendly.  

> If you add an image to the existing 100 images, you have to iterate training using batch sampling among the 101 images usually.

It should be theoretically possible to take a model that was trained with 100 images and then train it on a new image.  This would give you a model that was trained on 101 images.  I haven't actually tried this approach, but it seems similar to training all 101 images from scratch with a batch size of 1.  There may be more nuance here related to [iid training data](https://www.quora.com/Does-the-order-of-training-data-matter-when-training-neural-networks/answer/Ian-Goodfellow) than I have suggested, but even if the training happens from scratch every time with the current coreml workflow, I still think incorporating this training into the tool could be beneficial for speeding up workflows.  You could allow users to select how often the model is retrained(every 2 images or 20 images, etc.) and training could happen in the background so that the user could continue labeling while the model is training.

> What do you mean by "model based data exploration"?
> Is it the image sorting using the confidence scores?

Yeah, that would be an example.  You could also use the model to identify potential duplicates(although there are other approaches to this that don't require a model).  I'm not sure of how relevant heatmaps are to object detection, but that could be another avenue of model based data vis and there may be others that haven't come to mind.

**Comment by ryouchinsa** [comment]
On VoTT, an active learning means detect boxes using a trained model.
https://github.com/microsoft/VoTT

On RectLabel, we have "Process the image using Core ML model" feature.

We understand you want "Training a model in the background" feature.

**Comment by ryouchinsa** [comment]
Currently If you need our support to fix this problem, please let us know.
